feeadjuster/feeadjuster.py

```diff
@@ -0,0 +1,122 @@
+#!/usr/bin/env python3
+import random
+import time
+from pyln.client import Plugin, Millisatoshi, RpcError
+
+
+plugin = Plugin()
+# Our amount and the total amount in each of our channel, indexed by scid
+plugin.adj_balances = {}
+# Cache to avoid loads of calls to getinfo
+plugin.our_node_id = None
+
+
+def get_ratio(our_percentage):
+    """
+    Basic algorithm: the farther we are from the optimal case, the more we
+    bump/lower.
+    """
+    return 50**(0.5 - our_percentage)
+
+
+def maybe_adjust_fees(plugin: Plugin, scids: list):
+    for scid in scids:
+        our = plugin.adj_balances[scid]["our"]
+        total = plugin.adj_balances[scid]["total"]
+        percentage = our / total
+        last_percentage = plugin.adj_balances[scid].get("last_percentage")
+
+        # Only update on substantial balance moves to avoid flooding, and add
+        # some pseudo-randomness to avoid too easy channel balance probing
+        update_treshold = 0.05
+        if not plugin.deactivate_fuzz:
+            update_treshold += random.uniform(-0.015, 0.015)
+
+        if (last_percentage is None
+                or abs(last_percentage - percentage) > update_treshold):
+            ratio = get_ratio(percentage)
+            try:
+                plugin.rpc.setchannelfee(scid, int(plugin.adj_basefee * ratio),
+                                         int(plugin.adj_ppmfee * ratio))
+                plugin.log("Adjusted fees of {} with a ratio of {}"
+                           .format(scid, ratio))
+                plugin.adj_balances[scid]["last_percentage"] = percentage
+            except RpcError as e:
+                plugin.log("Could not adjust fees for channel {}: '{}'"
+                           .format(scid, e), level="warn")
+
+
+def get_chan(plugin: Plugin, scid: str):
+    for peer in plugin.rpc.listpeers()["peers"]:
+        if len(peer["channels"]) == 0:
+            continue
+        chan = peer["channels"][0]
+        if "short_channel_id" not in chan:
+            continue
+        if chan["short_channel_id"] == scid:
+            return chan
+
+
+def maybe_add_new_balances(plugin: Plugin, scids: list):
+    for scid in scids:
+        if scid not in plugin.adj_balances:
+            # At this point we could call listchannels and pass the scid as
+            # argument, and deduce the peer id (!our_id). But -as unlikely as
+            # it is- we may not find it in our gossip (eg some corruption of
+            # the gossip_store occured just before the forwarding event).
+            # However, it MUST be present in a listpeers() entry.
+            chan = get_chan(plugin, scid)
+            assert chan is not None
```

Comment:
It just failed :sob:

Reply:
Wired. Did a channel recently close?

Reply:
Even if it recently closed it'd be still present in listpeers, going to add debug lines and hope for the best..
It did not happen to you ?

Reply:
Nah

Reply:
It can happen if scid changes on reorg maybe

Reply:
Maybe indexing list element `0` failed?:  `chan = peer["channels"][0]`
Do you have a stracktrace?

Reply:
Maybe it was a peer where you had a prior channel (`0`) but the scid was on a newer channel with the same peer..  :D

Reply:
> It can happen if scid changes on reorg maybe

:sweat_smile: 

> Do you have a stracktrace?

```
2020-10-15T18:46:58.816Z INFO    plugin-feeadjuster.py: Traceback (most recent call last):
2020-10-15T18:46:58.816Z INFO    plugin-feeadjuster.py:   File \"/home/bitcoin/.local/lib/python3.8/site-packages/pyln/client/plugin.py\", line 497, in _dispatch_notification
2020-10-15T18:46:58.816Z INFO    plugin-feeadjuster.py:     self._exec_func(func, request)
2020-10-15T18:46:58.816Z INFO    plugin-feeadjuster.py:   File \"/home/bitcoin/.local/lib/python3.8/site-packages/pyln/client/plugin.py\", line 455, in _exec_func
2020-10-15T18:46:58.816Z INFO    plugin-feeadjuster.py:     return func(*ba.args, **ba.kwargs)
2020-10-15T18:46:58.816Z INFO    plugin-feeadjuster.py:   File \"/home/bitcoin/.lightning/plugins/feeadjuster/feeadjuster.py\", line 82, in forward_event
2020-10-15T18:46:58.816Z INFO    plugin-feeadjuster.py:     maybe_add_new_balances(plugin, [in_scid, out_scid])
2020-10-15T18:46:58.816Z INFO    plugin-feeadjuster.py:   File \"/home/bitcoin/.lightning/plugins/feeadjuster/feeadjuster.py\", line 69, in maybe_add_new_balances
2020-10-15T18:46:58.816Z INFO    plugin-feeadjuster.py:     assert chan is not None
2020-10-15T18:46:58.816Z INFO    plugin-feeadjuster.py: AssertionError
```

> Maybe it was a peer where you had a prior channel (0) but the scid was on a newer channel with the same peer.. :D

Ooooh now that's interesting :D

Reply:
Closing/Resolving due to IRC discussion...